### DISCUSSION TITLE
How to run controller code in debugger?

### DISCUSSION THREAD

**Discussion by olavt** [zwave-js/zwave-js#5231] [Q&A] [answered]
I'm new to Typescript / Node.

What are the minimum steps to get a controller up and running in a debugger (using either VS Code or VS 2022)? I just want to test including a new node. I have a Silicon Labs UZB7 connected to a Windows 11 computer. I have installed Node / Typescript and I can run simple console apps from VS 2022, but not sure on how to run the Z-Wave JS project.

**Comment by AlCalzone** [comment, marked as answer]
The minimum would be the `run.ts` script in the `test` folder which I use for debugging. There's a debug task "Debug locally" (or sth like that) you can execute in VSCode.

**Reply by olavt** [reply]
Thanks! A quick follow-up question: How would I be able to set the controller into inclusion-mode? How do I communicate with the controller? How do I run a minimal kind of UI or commands against the controller (while it's running in a debugger)?

**Reply by AlCalzone** [reply]
In the script you'll have to write code to interact with the driver, see documentation for available commands.

If you need a UI, I recommend using https://github.com/zwave-js/zwave-js-ui
You can open that in vscode and debug the driver by executing `yarn run dev:server` in a JavaScript debug terminal. @robertsLando may help you with questions about that project.

**Reply by olavt** [reply]
How do I fix this?

<img width="527" alt="image" src="https://user-images.githubusercontent.com/3481146/204791927-abbbdb94-20c9-4a73-a323-055756617674.png">

**Reply by olavt** [reply]
I don't actually need a full UI. Basically, I just need to send the driver a simple command to get the controller into inclusion mode (ADD_NODE_START).

**Reply by olavt** [reply]
"In the script you'll have to write code to interact with the driver"

Are you refering to the block of code inside this block in run.ts?

node.once("ready", async () => {
...
		});

**Reply by AlCalzone** [reply]
About the dependency issue, you might want to read https://zwave-js.github.io/node-zwave-js/#/development/intro

Your code should go into the `once("driver ready"...` block. You can delete the rest in there.

**Reply by olavt** [reply]
I have it up and running now. Had to disable some eslint errors related to cr characters at the end of each line. Guess it's a Linux / Windows issue. Maybe the eslintrc.js should be fixed to avoid this in the first place? I also had to run "yarn run build" from the command prompt (after unblocking yarn.ps1), since VS Code otherwise complained that yarn.ps1 was unsigned (security issue). I now only miss the beginInclusion code in order to start the inclusion process. I have posted another question on that.
